### PR TITLE
[1.13] Run metrics tests on one node per role

### PR DESCRIPTION
**Pull request by gracedo**
## High-level description

Changes some metrics integration tests (where applicable) to run tests on one node per role (public/private agent, master).

## Corresponding DC/OS tickets (obligatory)

These DC/OS JIRA ticket(s) must be updated (ideally closed) in the moment this PR lands:

  - [DCOS_OSS-4356](https://jira.mesosphere.com/browse/DCOS_OSS-4356) Run metrics tests on one node from each role

## Checklist for all PRs

  - [x] Added a comprehensible changelog entry to `CHANGES.md` or explain why this is not a user-facing change: integration test only change
  - [x] Included a test which will fail if code is reverted but test is not. If there is no test please explain here:
  - [x] Read the [DC/OS contributing guidelines](https://github.com/dcos/dcos/blob/master/contributing.md)
  - [x] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)

### CODE REVIEW

**Comment by branden** [comment]
@mesosphere-mergebot label ready for review

File: packages/dcos-integration-test/extra/test_metrics.py [hunk 1]
```diff
@@ -219,6 +244,73 @@ def check_statsd_metrics():
             check_statsd_metrics()
 
 
+def test_metrics_public_agent_statsd(dcos_api_session):
```
Comment (branden) [inline packages/dcos-integration-test/extra/test_metrics.py]: The body of this test looks like a duplicate of `test_metrics_private_agent_statsd()`. Let's either merge those into the same test or extract the duplicated code into a helper function.

File: packages/dcos-integration-test/extra/test_metrics.py [hunk 2]
```diff
@@ -291,11 +383,6 @@ def check_metrics_metadata():
 
 @pytest.mark.skipif(expanded_config.get('security') == 'strict',
                     reason="Framework disabled for strict mode")
-@pytest.mark.xfailflake(
-    jira='DCOS_OSS-4568',
-    reason='Framework hello-world still running',
-    since='2018-12-14',
-)
```
Comment (branden) [inline packages/dcos-integration-test/extra/test_metrics.py]: Should removing this flake marker be in this PR? Seems out-of-scope.

File: packages/dcos-integration-test/extra/test_metrics.py [hunk 3]
```diff
@@ -79,85 +84,105 @@ def get_metrics_prom(dcos_api_session, node):
...
         def check_mesos_metrics():
-            response = get_metrics_prom(dcos_api_session, agent)
+            response = get_metrics_prom(dcos_api_session, dcos_api_session.slaves[0])
+            assert 'mesos_slave_uptime_secs' in response.text
+        check_mesos_metrics()
+
+    if dcos_api_session.public_slaves:
+        @retrying.retry(wait_fixed=STD_INTERVAL, stop_max_delay=METRICS_WAITTIME)
+        def check_mesos_metrics():
+            response = get_metrics_prom(dcos_api_session, dcos_api_session.public_slaves[0])
```
Comment (branden) [inline packages/dcos-integration-test/extra/test_metrics.py]: The bodies of these two `if` statements are identical. This test could instead collect the agents we intend to test into a list, and then run through that list in a `for` loop.

```
nodes = []
if dcos_api_session.slaves:
    nodes.append(dcos_api_session.slaves[0])
if dcos_api_session.public_slaves:
    nodes.append(dcos_api_session.public_slaves[0])

for node in nodes:
    @retrying.retry(wait_fixed=STD_INTERVAL, stop_max_delay=METRICS_WAITTIME)
    def check_mesos_metrics():
        response = get_metrics_prom(dcos_api_session, node)
        assert 'mesos_slave_uptime_secs' in response.text
    check_mesos_metrics()
```

And this is doing the same exact thing as `test_metrics_masters_mesos()`. We could test a master here as well.

File: packages/dcos-integration-test/extra/test_metrics.py [hunk 4]
```diff
@@ -79,85 +84,105 @@ def get_metrics_prom(dcos_api_session, node):
...
 
-def test_metrics_agents_statsd(dcos_api_session):
-    """Assert that statsd metrics on agent are present."""
-    if len(dcos_api_session.slaves) > 0:
+        @retrying.retry(wait_fixed=STD_INTERVAL, stop_max_delay=METRICS_WAITTIME)
+        def check_adminrouter_metrics():
+            response = get_metrics_prom(dcos_api_session, dcos_api_session.public_slaves[0])
+            for metric_name in expected_metrics:
+                assert metric_name in response.text
+        check_adminrouter_metrics()
```
Comment (branden) [inline packages/dcos-integration-test/extra/test_metrics.py]: This `if` statement is copied from the above `if`; let's refactor so there isn't any code duplication.

File: packages/dcos-integration-test/extra/test_metrics.py [hunk 5]
```diff
@@ -28,39 +28,44 @@ def check_tags(tags: dict, expected_tag_names: set):
...
-    for agent in dcos_api_session.slaves:
-        response = dcos_api_session.session.request('GET', 'http://' + agent + ':61091/metrics')
+    if dcos_api_session.slaves:
+        response = dcos_api_session.session.request('GET', 'http://' + dcos_api_session.slaves[0] + ':61091/metrics')
+        assert response.status_code == 200, 'Status code: {}'.format(response.status_code)
+
+    if dcos_api_session.public_slaves:
+        response = dcos_api_session.session.request(
+            'GET', 'http://' + dcos_api_session.public_slaves[0] + ':61091/metrics')
         assert response.status_code == 200, 'Status code: {}'.format(response.status_code)
```
Comment (branden) [inline packages/dcos-integration-test/extra/test_metrics.py]: More duplicated code here. This is another test where it probably makes sense to merge with its corresponding `test_metrics_masters_prom()`. They're all duplicating the same assertion logic.

**Comment by branden** [comment]
@mesosphere-mergebot label ready for review

**Comment by gracedo** [comment]
@mesosphere-mergebot label Ready For Review

**Comment by gracedo** [comment]
@branden would you be able to re-review this when you have the chance? thanks :)

File: packages/dcos-integration-test/extra/test_metrics.py [hunk 6]
```diff
@@ -361,27 +367,14 @@ def wait_for_node_response(node):
...
+    if dcos_api_session.slaves:
+        nodes.append(dcos_api_session.slaves[0])
+    if dcos_api_session.public_slaves:
+        nodes.append(dcos_api_session.public_slaves[0])
 
-    # masters
-    for master in dcos_api_session.masters:
-        response = wait_for_node_response(master)
+    for node in nodes:
+        response = wait_for_node_response(node)
```
[the remark this answers is not in the document]
Comment (gracedo) [reply packages/dcos-integration-test/extra/test_metrics.py]: it's collapsed in the view, but right underneath this assertion https://github.com/dcos/dcos/pull/4077/files#diff-a99db445cccf205fe11239e4be6de2aaR381

[the remark this answers is not in the document]
Comment (branden) [reply packages/dcos-integration-test/extra/test_metrics.py]: Oh, my mistake. 👍

**Pull request merged**